pkg/controller/perconaservermongodbrestore/physical.go

```diff
@@ -187,171 +191,319 @@ func (r *ReconcilePerconaServerMongoDBRestore) reconcilePhysicalRestore(
+			return err
+		}
+		if !sts.DeletionTimestamp.IsZero() {
+			return nil
+		}
+
+		if sts.Labels[naming.LabelKubernetesComponent] != naming.ComponentArbiter {
+			if !hasContainerName(sts.Spec.Template.Spec.Containers, naming.ContainerBackupAgent) {
+				return errors.Errorf("statefulset %s wasn't deleted", sts.Name)
 			}
+		}
+
+		if sts.Annotations[psmdbv1.AnnotationRestoreInProgress] == "true" {
+			return nil
+		}
+
+		stsIsUpdated = false
 
-			if err := r.client.Delete(ctx, &sts); err != nil {
-				return status, errors.Wrapf(err, "delete statefulset %s", stsName)
+		return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			if err := r.client.Get(ctx, client.ObjectKeyFromObject(s), sts); err != nil {
+				return err
 			}
 
-			if rs.NonVoting.Enabled {
-				stsName := naming.NonVotingStatefulSetName(cluster, rs)
+			sts.Annotations[psmdbv1.AnnotationRestoreInProgress] = "true"
 
-				log.Info("Deleting statefulset", "statefulset", stsName)
+			return r.client.Update(ctx, sts)
+		})
+	}); client.IgnoreNotFound(err) != nil {
+		return false, errors.Wrap(err, "delete restore in progress annotation")
+	}
+	if !stsIsUpdated {
+		return false, nil
+	}
 
-				sts := appsv1.StatefulSet{
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      stsName,
-						Namespace: cluster.Namespace,
-					},
-				}
+	ready := true
+	err := r.iterateOverMongodSts(ctx, cluster, func(s *appsv1.StatefulSet) error {
+		if !ready {
+			return nil
+		}
 
-				if err := r.client.Delete(ctx, &sts); err != nil {
-					return status, errors.Wrapf(err, "delete statefulset %s", stsName)
+		var err error
+		ready, err = r.isStatefulSetReady(ctx, cluster, s)
+		return err
+	})
+	if client.IgnoreNotFound(err) != nil {
+		return false, err
+	}
+	if !ready || k8serrors.IsNotFound(err) {
+		return false, err
+	}
+
+	wait := false
+	if err := r.iterateOverMongodSts(ctx, cluster, func(s *appsv1.StatefulSet) error {
+		if s.Labels[naming.LabelKubernetesComponent] != naming.ComponentMongod {
+			return nil
+		}
+		rs := cluster.Spec.Replset(s.Labels[naming.LabelKubernetesReplset])
+		c, err := r.MongoClient().Mongo(ctx, cluster, rs, api.RoleClusterAdmin)
+		if err != nil {
+			wait = true
+
+			if errors.Is(err, topology.ErrServerSelectionTimeout) && strings.Contains(err.Error(), "ReplicaSetNoPrimary") {
+				pods, err := r.getReplsetPods(ctx, cluster, rs, s.Labels[naming.LabelKubernetesComponent])
+				if err != nil {
+					return err
+				}
+				updateConfig := func(pod corev1.Pod) error {
+					cli, err := r.MongoClient().Standalone(ctx, cluster, rs, api.RoleClusterAdmin, pod)
+					if err != nil {
+						return nil
+					}
+					defer func() { _ = cli.Disconnect(ctx) }()
+
+					cfg, err := cli.ReadConfig(ctx)
+					if err != nil {
+						return errors.Wrap(err, "read replset config")
+					}
+
+					if err := cli.WriteConfig(ctx, cfg, true); err != nil {
+						return errors.Wrap(err, "reconfigure replset")
+					}
```

Comment:
i don't get what we are doing here, we are reconfiguring replset but we are not changing anything in config?